Get values of grouped columns

jlumpe opened the pull request:
I noticed there doesn't seem to be a built-in mechanism to get the actual group values from a `GroupedDataFrame`, so I added a `groupvalues()` function:

```julia
julia> df = DataFrame(a = repeat([:foo, :bar, :baz], outer=[2]),
                      b = repeat([2, 1], outer=[3]),
                      c = 1:6);

julia> gd = groupby(df, :a)
GroupedDataFrame with 3 groups based on key: a
First Group (2 rows): a = :foo
│ Row │ a      │ b     │ c     │
│     │ Symbol │ Int64 │ Int64 │
├─────┼────────┼───────┼───────┤
│ 1   │ foo    │ 2     │ 1     │
│ 2   │ foo    │ 1     │ 4     │
⋮
Last Group (2 rows): a = :baz
│ Row │ a      │ b     │ c     │
│     │ Symbol │ Int64 │ Int64 │
├─────┼────────┼───────┼───────┤
│ 1   │ baz    │ 2     │ 3     │
│ 2   │ baz    │ 1     │ 6     │

julia> groupvalues(gd)
3-element Array{Tuple{Symbol},1}:
 (:foo,)
 (:bar,)
 (:baz,)
```

Unfortunately, `GroupedDataFrame` doesn't seem to remember whether it was given a single column or a vector containing a single column, so this will always return a vector of tuples.

In Pandas, iterating over the grouped object gives `(value, dataframe)` pairs, which I use very frequently. Of course you can always get the value from the first row of the `SubDataFrame` when iterating, but I think this feature leads to much cleaner code. I defined `Base.pairs(::GroupedDataFrame)` for this purpose:

```julia
julia > for ((a,), group) in pairs(groupby(df, :a))
     println("Process group $a, $(size(df, 1)) rows.")
end
Process group foo, 2 rows.
Process group bar, 2 rows.
Process group baz, 2 rows.
```

bkamins commented:
Thank you for the commit. If we should add `pairs` method depends on what @nalimilan plans to do with `GroupedDataFrame` object in the future.

bkamins commented on src/DataFrames.jl:
```diff
@@ -40,6 +40,7 @@ export AbstractDataFrame,
        eltypes,
        groupby,
        groupindices,
+       groupvalues,
```
also please add an entry to docs/src/lib/functions.md

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -1214,3 +1214,48 @@ groupindices(gd::GroupedDataFrame) = replace(gd.groups, 0=>missing)
+ (:bar,)
+ (:baz,)
+"""
+groupvalues(gd::GroupedDataFrame) = map(Tuple, eachrow(gb.parent[gb.idx[gb.starts], gb.cols]))
```
I think we should rather return `NamedTuples`:
```suggestion
groupvalues(gd::GroupedDataFrame) = [copy(sdf[1, gd.cols]) for sdf in gd]
```
what do you think?

bkamins replied:
can you also please later add tests to this?

bkamins replied:
given @nalimilan suggestion maybe we should return a `DataFrame`, so essentially it would be:
```
groupvalues(gd::GroupedDataFrame) = combine(sdf -> sdf[1, gd.cols], gd)
```

bkamins commented:
@nalimilan - maybe we could even consider storing the group infromation @jlumpe proposes when creating `GroupedDataFrame`? It should be cheap, and later we could reuse it.

nalimilan commented:
Thanks. I'm a bit hesitant about this API, since we currently never return arrays of (named) tuples: we rather use data frames for that. The `pairs` approach is appealing, but the problem is that indexing with values (if we added support for that, which would be required if `pairs` returns such values) is slower than using integer indices. Maybe we could define `eachgroup(gd, true)` to return `(groupvalues, sdf)` pairs, just like we have `eachcol(df, true)`.

bkamins commented:
Actually - if we followed this idea, we should add an argument to `groupby`. Then current behavior `groupby(df, some_columns)` is like `eachcol(df)` and the additional behavior `groupby(df, some_columns, true)` would be like `eachcol(df, true)`.

nalimilan commented:
It would be weird to have an argument to `groupby`, given that the grouping is the same: one would have to re-group just to change the iteration behavior.

I'm still not sure what's the best solution. The most appealing approach seems to be to define `keys` and `pairs` to return named tuples with the groups, and have `getindex` support that (for consistency). what do you think? People who want performance would use direct iteration or `eachindex`, which is defined to give the most efficient index type for arrays (as opposed to `keys`, so we would be consistent with that).

bkamins commented:
What we should do here depends if you want to make `GroupedDataFrame <: AbstractDataFrame` in the future. If not, then we are free to decide what is most convenient. However, if it is going to be a subtype of `AbstractDataFrame` we might have some constraints here.

nalimilan commented:
In the end I don't think we should make `GroupedDataFrame <: AbstractDataFrame`. Hadley said that he wouldn't do that if he were to design dplyr from scratch, and there aren't big advantages to doing that anyway (we can still define specific methods like `select` if we want). But I'm not sure the `AbstractDataFrame` interface would conflict here, since we're talking about indexing with a single value, which is now deprecated.

bkamins commented:
I think then it is OK to add `keys` and `pairs`. `getindex` could accept both integer indexing and named tuple indexing similar to how `NamedTuple` allows dual indexing.

jlumpe commented:
Implemented feedback:
* Added `groupvalues` to API docs.
* `groupvalues` now returns a `DataFrame`.
* Values now cached in a field of `GroupedDataFrame`.
* Implemented dictionary interface by defining methods for `getindex`, `keys` and `get`.
  * Keys are `NamedTuple`s and `getindex` supports `NamedTuple` keys with any field order or 
     regular `Tuples` that match the order of `gd.cols`.
  * `pairs` automatically works when `keys` is defined.
  * Did not make it a subtype of `AbstractDict`.

jlumpe commented:
Also, I thin this would resolve #1693

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -179,6 +206,43 @@ end
     GroupedDataFrame(gd.parent, gd.cols, gd.groups, gd.idx, gd.starts, gd.ends)
 
+function Base.keys(gd::GroupedDataFrame)
+    colnames = Tuple(_names(gd)[gd.cols])
```
maybe sipler is `_names(gd.values)`? Same below

bkamins replied:
Ah - and now I realize that if we materialize `values` then we run into the problem when `parent(gd)` columns are renamed. So probably it is safer not to store it, but compute dynamically (this adds a small overhead, but I assume that this function will not be used in performance-critical loops).

bkamins replied:
or we should store only a vector of tuples of values (as group values should not change after grouped data frame is created).

jlumpe replied:
Maybe just store the values as a tuple of column vectors? That seems to be the most direct way to create a DataFrame, and it can get the names from the parent each time it's called.

bkamins replied:
sounds OK

nalimilan replied:
Maybe we should return a custom iterator creating named tuples on the fly, rather than allocating a vector (that could just be a generator)? Also, better not store a `gd.values`, as it's going to add some overhead to all `groupby` operations even when it's not needed.

I was also thinking that iterator could itself return a special named tuple-like object which would internally keep track of the group number, so that we don't need to recompute if on indexing (which is terribly inefficient). What do you think?

bkamins replied:
Thinking of it now - I think we can be lazy here. The benefit of materializing is if we did many iterations over a `GroupedDataFrame` after it is created, which is unlikely. Alternatively (but I think this is not needed given the discussion), we could do something similar that we do for `SubIndex` - materialize it only if needed (i.e. create an empty store and populate it only if iteration over group vales is required by a user).

jlumpe replied:
Now stores `.values` as vector of unnamed tuples, `keys()` gets parent column names when called. Added test for parent `DataFrame` column names being changed after creation.

bkamins replied:
I unresolved it as I do not see these changes.

However (assuming that you have made changes but have not pushed them yet), what (I guess) @nalimilan and (for sure) I meant it that `values` should not be populated until it is being read for the first time.

bkamins replied:
OK - I am resolving it here and I left a new comment in the code review section.

bkamins commented on src/groupeddataframe/grouping.jl:
````diff
@@ -1214,3 +1278,47 @@ groupindices(gd::GroupedDataFrame) = replace(gd.groups, 0=>missing)
+│ 3   │ baz    │
+```
+"""
+groupvalues(gd::GroupedDataFrame) = gd.values
````
If we return here a `DataFrame` (no matter what we store in `values` and if we store anything) then above in indexing we should allow `DataFrameRow` in `getindex` also (as it is a mutable dual of a `NamedTuple`).

But I am not sure if we need this function at all, as it serves the same purpose as `keys` (and `keys` returns `NamedTuples`, so then we do not need to support `DataFrameRow` in `getinex`).

jlumpe replied:
Maybe just store the values as a tuple of column vectors? That seems to be the most direct way to create a DataFrame, and it can get the names from the parent each time it's called.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -63,6 +69,12 @@ See the following for additional split-apply-combine operations:
 
+`GroupedDataFrame` also supports the dictionary interface, where its keys are
+`NamedTuple`s containing the values of the grouped columns. When indexing using
+a `NamedTuple` key the fields may be in any order. You may also index using a
```
do you find "any order" a valuable feature here. I usually tend to think that we should require a strict order (but opinions are mixed here 😄).

jlumpe replied:
I'm not sure if its a super valuable feature, but it was easy enough to code and I don't think creates too much overhead.

bkamins replied:
We are strict not because of overhead but safety (i.e. if one passes column names in a wrong order then we are not sure if one would want to get a result or rather an error).

nalimilan replied:
I agree, better be strict. If we find out it's really annoying in practice we can be more tolerant in the future, but let's keep things simple and strict for now.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -179,6 +206,43 @@ end
 Base.getindex(gd::GroupedDataFrame, idxs::Colon) =
     GroupedDataFrame(gd.parent, gd.cols, gd.groups, gd.idx, gd.starts, gd.ends)
 
+function Base.keys(gd::GroupedDataFrame)
```
I would add a docstring to `keys` and `get`.

bkamins commented on docs/src/lib/functions.md:
```diff
@@ -16,6 +16,7 @@ by
 combine
 groupby
 groupindices
+groupvalues
```
Thank you for adding this. Also it would be great to add an entry to docs/src/lib/indexing.md with `GroupedDataFrame` section for `getindex` as now it becomes complex enough so that it should be precisely documented.

nalimilan replied:
Can you drop this function for now? If we define `keys`, it's probably not necessary anymore, right?

bkamins replied:
this is what I also suggested.

jlumpe replied:
Dropped it.

bkamins commented:
Thank you for the fixes. I have left some things that should be discussed (especially with @nalimilan feedback) - so probably it is best to finish the design before you implement the changes.

jlumpe commented:
Implemented your feedback, this is now essentially just implementing the dictionary interface for `GroupedDataFrame`:

* Store grouped values as vector of plain tuples in field `.values`.
* `keys()` now looks up the parent column names when called, so it changes if the parent's columns are renamed (added a test for this).
* Out-of-order `NamedTuple` keys no longer supported. Plain `Tuple` keys in correct order are supported still.
* `groupvalues` function has been removed, `DataFrame(keys(td))` now achieves the same thing.
* Added explanation of `GroupedDataFrame` indexing to docs
* Also added tip about using `pairs()` function to get group values when iterating.

bkamins commented on docs/src/lib/indexing.md:
```diff
@@ -189,3 +189,15 @@ Note that `sdf[!, col] .= v` and `sdf[!, cols] .= v` syntaxes are not allowed as
+
+[`GroupedDataFrame`](@ref) implements the dictionary interface and so supports additional methods for indexing:
+
+* `gd[i::Int]` -> Get the `i`th group
```
```suggestion
* `gd[i::Integer]` -> Get the `i`th group
```

bkamins commented on docs/src/man/split_apply_combine.md:
````diff
@@ -138,3 +138,15 @@ julia> for subdf in groupby(iris, :Species)
 50
 ```
+
+To also get the values of the grouped columns along with each group, use the
````
@nalimilan - you are the master of docstring reviews :), but should we say here "grouped columns" or "grouping columns"? Also maybe we can mention that the values are stored in a `NamedTuple`?

nalimilan replied:
Yes, "grouping columns" sounds better. Mentioning `NamedTuple` also sounds like a good idea.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -8,11 +8,17 @@ Not meant to be constructed directly, see `groupby`.
+    idx::Vector{Int}          # indexing vector when grouped by the given columns
+    starts::Vector{Int}       # starts of groups
+    ends::Vector{Int}         # ends of groups
+    values::Vector{<: Tuple}  # values of grouped columns
```
```suggestion
    values::Vector{<:Tuple}  # values of grouped columns
```

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -8,11 +8,17 @@ Not meant to be constructed directly, see `groupby`.
+end
+
+function GroupedDataFrame(parent, cols, groups, idx, starts, ends)
+    values = [Tuple(parent[idx[s], cols]) for s in starts]
```
I will return to an earlier discussion. As @nalimilan suggested - this can be very expensive if we have a lot of small groups. Maybe for the time being we can leave-out the `values` field and only calculate it on the fly (which should not be problematic unless we reuse `gd` many times for iteration later, which I guess is a rare case, and if someone really wants it you can always store `keys` to get this)?

The problem with lazy creation of `values` field is that we cannot do it without materializing as `Tuple`s inside it might have heterogeneous type.

bkamins replied:
Also - if we decide to materialize `values` probably `Vector` is not optimal, as it would be better to use `Dict`. The reason is that `keys` is an operation that you do only once (and then `Vector` is useful), but `findfirst` operation is `O(n)`, so in practice the lookup in `getindex` will be slow with `Vector`.

Finally, no matter if we use `Tuple` or `Dict` unless we improve the code here we will normally get a bad type inference in corner cases with `Missing` (which are to be expected in DataFrames.jl). What I mean is:
```
julia> [(1,missing), (missing, 1), (1,2)]
3-element Array{Tuple{Any,Any},1}:
 (1, missing)
 (missing, 1)
 (1, 2)
```
and
```
julia> Dict((1,missing)=>1, (missing, 1)=>2, (1,2)=>3)
Dict{Tuple{Any,Any},Int64} with 3 entries:
  (1, 2)       => 3
  (missing, 1) => 2
  (1, missing) => 1
```

In summary, we can do two things:
1. leave out materializing `values` and warn users that the `getindex` and `keys` operation is currently lazy and might be slow in hot loops (so they should themselves materialize the groups and use integer indexing which is fast for such use-cases)
2. work out the best design of `values` related field (that will not have performance impact if this filed is not needed, has correct eltype identification and is fast in the relevant use-cases)

I would opt for 1 in this PR (as this is functionality related) and leave option 2 for a separate PR (for @jlumpe to submit - if you wish; alternatively I can think about it separately). The reason is that then we will already have a working functionality in DataFrames.jl and only performance optimization (which is non-breaking, but potentially it will be a longer discussion about what is the best design here), will be left for later.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -66,6 +72,11 @@ See the following for additional split-apply-combine operations:
 
+`GroupedDataFrame` also supports the dictionary interface, where its keys are
+`NamedTuple`s containing the values of the grouped columns. You may also index
+using a `Tuple`, in which case the values should be in the same order as the `cols`
```
`Tuple` indexing is non-standard, but I am OK to allow it as it might be convenient. @nalimilan - any thoughts on this?

nalimilan replied:
I'm not sure there's a strong use case for it either, but it doesn't sound too problematic, so...

bkamins replied:
I am OK with `Tuple` now.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -182,6 +201,128 @@ end
+end
+
+# Get group by values of grouped column (named Tuple)
+function Base.getindex(gd::GroupedDataFrame, key::NamedTuple{N}) where N
```
```suggestion
function Base.getindex(gd::GroupedDataFrame, key::NamedTuple{N}) where {N}
```

bkamins commented:
@jlumpe - thank you. I have left several minor comments and one major one regarding `values` field. My current thinking is that we can leave-out the decision if and how we implement it for a separate PR as this is only performance optimization issue (and how to best design it is non-obvious so it will be easier to discuss it when we have the core user-facing functionality already merged to master).

nalimilan commented on docs/src/lib/indexing.md:
```diff
@@ -189,3 +189,15 @@ Note that `sdf[!, col] .= v` and `sdf[!, cols] .= v` syntaxes are not allowed as
+
+## Indexing `GroupedDataFrame`s
+
+[`GroupedDataFrame`](@ref) implements the dictionary interface and so supports additional methods for indexing:
```
```suggestion
[`GroupedDataFrame`](@ref) implements the `AbstractDict` interface and so supports additional methods for indexing:
```

nalimilan commented on docs/src/lib/indexing.md:
```diff
@@ -189,3 +189,15 @@ Note that `sdf[!, col] .= v` and `sdf[!, cols] .= v` syntaxes are not allowed as
+
+* `gd[i::Int]` -> Get the `i`th group
+* `gd[key::NamedTuple]` -> Get the group corresponding to the dictionary key `key` (see [`keys(::GroupedDataFrame)`](@ref)).
+  The fields of the `NamedTuple` must be in the same order, which is the order of the columns passed to [`groupby`](@ref).
```
```suggestion
  The fields of the `NamedTuple` must match (including order) the grouping columns passed to [`groupby`](@ref).
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -182,6 +201,128 @@ end
     GroupedDataFrame(gd.parent, gd.cols, gd.groups, gd.idx, gd.starts, gd.ends)
 
+"""
+    Base.keys(gd::GroupedDataFrame)
```
```suggestion
    keys(gd::GroupedDataFrame)
```

nalimilan commented on docs/src/man/split_apply_combine.md:
````diff
@@ -138,3 +138,15 @@ julia> for subdf in groupby(iris, :Species)
+
+```jldoctest sac
+julia> for (key, subdf) in pairs(groupby(iris, :Species))
+           println("Number of data points for $(key.Species): $(size(subdf, 1))")
````
```suggestion
           println("Number of data points for $(key.Species): $(nrow(subdf))")
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -182,6 +201,128 @@ end
+end
+
+"""
+    Base.get(gd::GroupedDataFrame, key, default)
```
```suggestion
    get(gd::GroupedDataFrame, key, default)
```

bkamins commented:
@jlumpe - I would love to see this PR polished and merged before release 1.0 (which we will have this year probably). How do you see this?

jlumpe commented:
Ok, changed it so it calculates the keys/values on the fly instead of storing them in a field, and implemented the suggested changes to documentation/docstrings.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -178,9 +191,135 @@ function Base.getindex(gd::GroupedDataFrame, idxs::AbstractArray)
+"""
+function Base.keys(gd::GroupedDataFrame)
+    colnames = Tuple(_names(gd)[gd.cols])
+    coltypes = [eltype(gd.parent[:, c]) for c in gd.cols]
```
```suggestion
    coltypes = [eltype(gd.parent[!, c]) for c in gd.cols]
```

we do not need to make a copy here. Right?

jlumpe replied:
Is that what the `!` syntax does?

bkamins replied:
`df[:, c]` copies the column and `df[!, c]` does not copy it (i.e. exposes column as it is stored in the `DataFrame` - in general it is unsafe, but here we just want to check its type and we do not modify it so it is OK, and it is much cheaper)

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -178,9 +191,135 @@ function Base.getindex(gd::GroupedDataFrame, idxs::AbstractArray)
+
+# Get group by values of grouped column (plain Tuple)
+function Base.getindex(gd::GroupedDataFrame, key::Tuple)
+    for (i, v) in enumerate(_groupvalues(gd))
```
It is OK as is, but we could do it without materializing `_groupvalues(gd)` which would be a bit less expensive (but we can leave it as is and do what I propose later as an optimization - it will be relevant only for `GroupedDataFrame` with a lot of groups).

jlumpe replied:
Here `_groupvalues()` returns a generator, so it will not allocate the entire array up front.

bkamins replied:
Ah - sorry. Then we have an issue that generator has `eltype` `Any` so it will not be super fast. But as I have said earlier I think we can leave it as is (otherwise we could test the approach using the function barrier to speed it up, but it does not guarantee to be faster so I think it is OK to leave it as is).

nalimilan replied:
AFAICT you can just remove the single-argument `_groupvalues` method and do this here instead. That sounds clearer:
```suggestion
    for i in 1:length(gd)
        v = _groupvalues(gd, i)
```

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -178,9 +191,135 @@ function Base.getindex(gd::GroupedDataFrame, idxs::AbstractArray)
+
+# Get group by values of grouped column (named Tuple)
+function Base.getindex(gd::GroupedDataFrame, key::NamedTuple{N}) where {N}
+    if length(key) != length(gd.cols) || N != Tuple(_names(gd)[gd.cols])
```
Again `_names(gd)[gd.cols]` performs an unnecessary allocation (also the same pattern is used above), so it could be optimized, but for now we can leave it as is and optimize the code laterr.

bkamins commented:
Thank you! Looks good to me for now (I left some performance notes, but they can be left for later if you prefer).

Let us just wait for @nalimilan to have a look at the PR.

jlumpe commented:
Implemented your suggestions.

nalimilan commented:
How about returning a generator of custom objects instead of a vector of named tuples (as proposed at https://github.com/JuliaData/DataFrames.jl/pull/1908#discussion_r312750043)? These objects would behave like named tuples (i.e. they would implement `keys`, `propertynames`, `getindex` and `getproperty`), but would also store the integer index of the group. That way indexing using these would be fast, instead of being a performance trap.

jlumpe commented:
@nalimilan that's an interesting idea. I guess it's probably true that users will be using the key objects returned by `keys()` more often than constructing them themselves. In that case we should probably keep the current `NamedTuple` indexing too, right?

nalimilan commented:
Yes, indexing with named tuples should still be allowed for convenience.

asinghvi17 commented:
Bump?

bkamins commented:
@jlumpe Given the bump by @asinghvi17 can you please let us know how you see the comments by @nalimilan and by me, so that we can finalize this PR? Thank you!

bkamins commented:
Nice approach. I will have a look at the details during the weekend. Could you share some benchmark results?

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +194,193 @@ end
+"""
+Key for one of the groups of a [`GroupedDataFrame`](@ref). Contains the values
+of the corresponding grouping columns and behaves similarly to a `NamedTuple`,
+but using it to index its GroupedDataFrame is much more effecient than using the
```
```suggestion
but using it to index its `GroupedDataFrame` is much more effecient than using the
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +194,193 @@ end
 
+
+"""
+Key for one of the groups of a [`GroupedDataFrame`](@ref). Contains the values
```
```suggestion
    GroupKey

Key for one of the groups of a [`GroupedDataFrame`](@ref). Contains the values
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +194,193 @@ end
+but using it to index its GroupedDataFrame is much more effecient than using the
+equivalent `Tuple` or `NamedTuple`.
+
+Instances of this type are returned by [`Base.keys(::GroupedDataFrame)`](@ref)
```
```suggestion
Instances of this type are returned by [`keys(::GroupedDataFrame)`](@ref)
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +194,193 @@ end
+function Base.keys(gd::GroupedDataFrame)
+    N = Tuple(_names(gd)[i] for i in gd.cols)
+    V = Tuple{_grouptypes(gd)...}
+    return [GroupKey{N, V}(gd, i, values) for (i, values) in enumerate(_groupvalues(gd))]
```
Better return a generator, allocating a full array is going to be quite costly. Also, why generate `values` upfront? We can defer this until `getproperty` is called on the `GroupKey` (which may never happen).

jlumpe replied:
I have made value access lazy, do you think it is still necessary to return a generator here?

nalimilan replied:
Yes, it makes even more sense, so that everything is lazy.

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +194,193 @@ end
+"""
+    keys(gd::GroupedDataFrame)
+
+Get a set of keys for each group of the `GroupedDataFrame`. The keys can be used
```
I suggest something like this:
```
Get the set of keys for each group in `GroupedDataFrame` object `gd`.
The keys are [`GroupKey`](@ref)s objects which behave like named tuples holding
the values of the grouping keys and can be used to index into `gd` efficiently.
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +194,193 @@ end
+    return [GroupKey{N, V}(gd, i, values) for (i, values) in enumerate(_groupvalues(gd))]
+end
+
+# Get group by GroupKey
```
```suggestion
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +194,193 @@ end
+    throw(KeyError(NamedTuple(key)))
+end
+
+# Get group by values of grouped column (plain Tuple)
```
```suggestion
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +194,193 @@ end
+    throw(KeyError(key))
+end
+
+# Get group by values of grouped column (named Tuple)
```
```suggestion
```

nalimilan commented on test/grouping.jl:
```diff
@@ -1200,4 +1220,108 @@ end
+    @test eltype(gdkeys) == GroupKey{N, V}
+
+    # Check values
+    @test map(NamedTuple, gdkeys) == [(a=:A, b=:X), (a=:B, b=:X), (a=Missing, b=:X), (a=:A, b=:Y), (a=:B, b=:Y), (a=Missing, b=:Y)]
```
```suggestion
    @test map(NamedTuple, gdkeys) ≅
        [(a=:A, b=:X), (a=:B, b=:X), (a=missing, b=:X), (a=:A, b=:Y), (a=:B, b=:Y), (a=missing, b=:Y)]
```

nalimilan commented on test/grouping.jl:
```diff
@@ -1200,4 +1220,108 @@ end
+
+    @test names(gd) == names(df)
+    @test groupvars(gd) == [:a, :b]
+    @test map(NamedTuple, keys(gd)) ≅ [(a=:A, b=:X), (a=:B, b=:X), (a=missing, b=:X), (a=:A, b=:Y), (a=:B, b=:Y), (a=missing, b=:Y)]
```
```suggestion
    @test map(NamedTuple, keys(gd)) ≅
        [(a=:A, b=:X), (a=:B, b=:X), (a=missing, b=:X), (a=:A, b=:Y), (a=:B, b=:Y), (a=missing, b=:Y)]
```

nalimilan commented on test/grouping.jl:
```diff
@@ -1200,4 +1220,108 @@ end
+
+    @test names(gd) == names(df)
+    @test groupvars(gd) == [:d, :e]
+    @test map(NamedTuple, keys(gd)) ≅ [(d=:A, e=:X), (d=:B, e=:X), (d=missing, e=:X), (d=:A, e=:Y), (d=:B, e=:Y), (d=missing, e=:Y)]
```
```suggestion
    @test map(NamedTuple, keys(gd)) ≅
        [(d=:A, e=:X), (d=:B, e=:X), (d=missing, e=:X), (d=:A, e=:Y), (d=:B, e=:Y), (d=missing, e=:Y)]
```

jlumpe commented:
@bkamins What sort of data do you recommend I benchmark this on?

nalimilan commented on test/grouping.jl:
```diff
@@ -1218,4 +1238,126 @@ end
+
+    # Type stability for value access methods
+    @test Base.return_types(values, (K,)) == [V]
+    # _get1(x) = x[1]
```
?

jlumpe replied:
This is testing that the compiler can correctly infer that `values(::GroupKey{N,V})` returns the type V. Typically you would use `Test.@inferred` but it doesn't work here because V is `Tuple{Union{Missing, Symbol}, Symbol}` but the actual return type will be one of the two concrete subtypes of this.

bkamins replied:
I guess the question is why do you need the comment in line 1331.

jlumpe replied:
Ah, these were similar tests for accessing elements of the key by index or property name. These would also be inferred correctly for a NamedTuple, but I left them commented out because I haven't implemented that yet. I'm not sure if its worth it.

nalimilan commented:
> @bkamins What sort of data do you recommend I benchmark this on?

I guess some simple benchmarks will do. You can take inspiration from e.g. https://github.com/JuliaData/DataFrames.jl/pull/1520. What would be interesting is comparing the speed of a loop over groups using integer indices to that of a loop using `GroupKey`s.

bkamins commented on docs/src/lib/indexing.md:
```diff
@@ -197,3 +197,17 @@ Note that `sdf[!, col] .= v` and `sdf[!, cols] .= v` syntaxes are not allowed as
+* `gd[key::GroupKey]` -> Get the group corresponding to the [`GroupKey`](@ref)
+  `key` (see [`keys(::GroupedDataFrame)`](@ref)). This should be nearly as fast
+  as integer indexing.
+* `gd[key::NamedTuple]` -> Get a group given the values of the grouped columns.
```
```suggestion
* `gd[key::NamedTuple]` -> Get the group corresponding to the given values of the grouping columns.
```

bkamins commented on docs/src/lib/indexing.md:
```diff
@@ -197,3 +197,17 @@ Note that `sdf[!, col] .= v` and `sdf[!, cols] .= v` syntaxes are not allowed as
+* `gd[key::NamedTuple]` -> Get a group given the values of the grouped columns.
+  The fields of the `NamedTuple` must match the grouping columns columns passed to [`groupby`](@ref) (including order).
+* `gd[key::Tuple]` -> Same as previous, but omitting the names on `key`.
+* `get(gd, key, default)` -> Get group by key, returning `default` if it does not exist.
```
```suggestion
* `get(gd, key, default)` -> Get group by `key`, returning `default` if it does not exist.
```

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,242 @@ end
+end
+
+# Get values of grouping columns for single group by its index
+_groupvalues(gd::GroupedDataFrame, i::Integer) = Tuple(gd.parent[gd.idx[gd.starts[i]], gd.cols])
```
just to be sure all will work correctly in the future - maybe add another method for `i::Bool` that throws an error?

bkamins replied:
Ah - never mind - it will be a problem only on Julia 1.0, so probably we can ignore it.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,242 @@ end
+    i === nothing && throw(KeyError(n))
+    return key[i]
+end
+function Base.getproperty(key::GroupKey, p::Symbol)
```
also `propertynames` should be defined if we overload `getproperty`.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,242 @@ end
+
+Base.:(==)(k1::GroupKey, k2::GroupKey) = parent(k1) === parent(k2) && getfield(k1, :idx) == getfield(k2, :idx)
+
+Base.NamedTuple(key::GroupKey{N}) where {N} = NamedTuple{N}(values(key))
```
Looking at this line what is the benefit of your definition of `GroupKey` over simply defining it as:
```
struct GroupKey{T<:GroupedDataFrame}
    parent::T
    idx::Int
end
```

jlumpe replied:
Pretty much just the type inference you get from the extra type parameters.

bkamins replied:
The issue is (as commented below) if we later really get this type-inference benefit (so we pay a cost during creation to have type stability but the question is if we use this type stability later to get the benefits of paying this cost here).

As an extreme case have a look at e.g. https://discourse.julialang.org/t/bad-performance-of-group-by-of-dataframes/30061 from today that refers to `by` - in that use case it is actually better to have type unstable code.

I am not saying that what you do is bad - but we should just be sure that it is superior to what I proposed in a typical usage scenario as both have their pros and cons. Thank you for looking into this!

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,242 @@ end
+See [`keys(::GroupedDataFrame)`](@ref) for more information.
+"""
+struct GroupKey{N, T<:Tuple}
+    parent::GroupedDataFrame
```
`GroupedDataFrame` is a parametric type so it should be also moved as parameter to `GroupKey`.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,242 @@ end
+function Base.keys(gd::GroupedDataFrame)
+    N = Tuple(_names(gd)[i] for i in gd.cols)
+    V = Tuple{_grouptypes(gd)...}
+    return [GroupKey{N, V}(gd, i) for i in 1:length(gd)]
```
OK - maybe here is a benefit of making `GroupKey` parametric on `N` and `V` - right?

nalimilan replied:
Yes I guess in the case of a loop over `keys(gd)`, the compiler may be able to realize that `N` and `V` don't change across `GroupKey` objects and infer the type of the groups. But that needs to be checked.

bkamins replied:
I think you would get the benefit but you have to use a barrier function to get it. This is a speculation though.

@jlumpe - have you tested this? (probably using `@inferred` inside a barier function could be used in a test)

nalimilan replied:
Yes and for the function barrier to have an effect, `keys(gd)` would have to return an object parameterized on the column types. Probably not worth it though.

jlumpe replied:
@nalimilan The return type of `keys(gd)` is already parameterized by the column types?

bkamins replied:
Yes it is now, but if you create `keys` and do not use barier function then you get no benefit of it, because the compiler will not know these types at compile time anyway. So the question is (as posted above) - is it likely that users will write the code with barier function + even if they do then this barier function will need to be recompiled each time a new `groupby` is performed (due to type stability issues). So my concern is (and I assume also @nalimilan shares it) is this beneficial to have it in a typical usage scenario (for sure if you have millions of groups and use a barier function and within it you heavily use the return value of `keys` then there is a benefit, but the question is if this is a typical use case).

jlumpe replied:
Good question. I mostly did it because it seemed more correct and wasn't much extra work. I'm not sure if there will be much performance benefit as is, but I think that once a typed version of `DataFrame` is implemented this will need to be done anyways.

Now that I think of it, it might be beneficial to parameterize `GroupedDataFrame` instead. That should be much easier than for `DataFrame` and shouldn't be backwards incompatible. This information would then be available to `GroupKey` via the type parameter for its parent, which I just added.

nalimilan replied:
As long as `DataFrame` and `GroupedDataFrame` are not type-stable, probably better not parameterize the new types either unless you can see performance improvements in benchmarks. Parameters can actually make things slower if they are not inferred. These can always be added later if we make `DataFrame` type-stable at some point.

bkamins replied:
I agree. Also my intuition for the future is that type-stability in `DataFrame` will be optional (I am not sure how it will be achieved but this is a discussion for post 1.0 release).

jlumpe replied:
Makes sense, I removed all parameters but for the parent `GroupedDataFrame`.

jlumpe commented:
Removed the type parameters. I also more or less came full circle and promoted the private `_groupvalues` function back to a public one. There's some redundancy with `keys()` but it lets you directly get the values for a single group by its index, which would otherwise require instantiating all the `GroupKey`s.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+# Get grouping variable index by its name
+function _groupvar_idx(gd::GroupedDataFrame, name::Symbol, strict=true)
+    i = findfirst(==(name), groupvars(gd))
+    i === nothing && strict && error("$name is not a grouping column")
```
Could you please change `error` to `throw(ArgumentError( ...`. Thank you.

nalimilan replied:
Also this one. Maybe something went wrong?

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+groupvars(gd::GroupedDataFrame) = _names(gd)[gd.cols]
+
+# Get grouping variable index by its name
+function _groupvar_idx(gd::GroupedDataFrame, name::Symbol, strict=true)
```
```suggestion
function _groupvar_idx(gd::GroupedDataFrame, name::Symbol, strict::Bool=true)
```

nalimilan replied:
Still has to be applied. BTW, probably better not have a default value since it's internal: that way it's more explicit.

jlumpe replied:
Removed this function at some point

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+
+Base.parent(key::GroupKey) = getfield(key, :parent)
+Base.length(key::GroupKey) = length(parent(key).cols)
+Base.names(key::GroupKey) = Tuple(groupvars(parent(key)))
```
Why adding a method to `Base.names`? I think that `Base.keys` is enough. (`Base.names` is reserved for modules AFAIK).

bkamins replied:
OK - now I see why you added `names` 😄. I understand you want `GroupKey` behave like `DataFrameRow` here. Right? If so - then this is OK to leave it as is.

jlumpe replied:
I just did it for consistency with `DataFrame` and `GroupedDataFrame`, it's definitely not necessary because `keys()` does the same thing.

jlumpe replied:
I think I'll remove it, it's not defined for `NamedTuple` either.

bkamins replied:
This was my first reflection (i.e. that it is not defined for `NamedTuple`), but on the other hand it is defined for `DAtaFrameRow` (which also defines `keys`) so that is why I have given this additional comment that I think it is OK to leave `names` defined for `GrupKey`.

nalimilan replied:
Why return a tuple rather than a vector though? As long as the length isn't hardcoded in the type, that doesn't make anything faster.

jlumpe replied:
Makes sense, removed conversion to tuple.

jlumpe replied:
Makes sense, removed conversion to Tuple.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+Base.getindex(key::GroupKey, i::Integer) = groupvalues(parent(key), getfield(key, :idx), i)
+
+function Base.getindex(key::GroupKey, n::Symbol)
+    try
```
Why not add appropriate logic in https://github.com/JuliaData/DataFrames.jl/pull/1908/files#diff-51878bdcc3c7eb068f85c4e853b4554dR295 (i.e. check there if `n` is a grouping variable and throw `KeyError` if it is not? - you could avoid `try`-`catch` which slows things down here by throwing `KeyError` already in `_groupvaridx`), and here just call `groupvalues`?

jlumpe replied:
Sorry, I can't tell which part the link is referring to? `_groupvar_idx` returns `nothing` when the key isn't found and `strict=false`, so that could be used instead of a try/catch. Or if you think `KeyError` is the appropriate error type in all instances this function is used, I can just change the default and use it everywhere. l didn't realize try/catch was so expensive when no exception is thrown, but it seems like it is.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+
+function Base.getproperty(key::GroupKey, p::Symbol)
+    try
+        return key[p]
```
Similarly here - better make `key[p]` throw an appropriate error I think (and again - we avoid using `error` as it is too generic).

jlumpe replied:
I'm using `ErrorException` here to match the default behavior on an attempt to access a non-existent field, which happens with `NamedTuple`.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+    end
+end
+
+# Base.:(==)(k1::GroupKey, k2::GroupKey) = parent(k1) === parent(k2) && getfield(k1, :idx) == getfield(k2, :idx)
```
Why this line is commented out?

jlumpe replied:
Ah, should be removed. It turns out this matches up with what the default equality for structs is already.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+"""
+function Base.get(gd::GroupedDataFrame, key::Union{GroupKey, Tuple, NamedTuple}, default)
+    try
+        return gd[key]
```
Given this code we should make sure we throw `KeyError` if and only if `key` was not found in `gd` (I assume this is the case but please double check as you know the code best).

jlumpe replied:
That should be what it does (not in `groupvars(gd)`), what's the alternative?

bkamins replied:
I think this is what it does, but I simply meant to have a second look at it 😄.

jlumpe replied:
There's a test that checks for a KeyError in this case, so ti should be good.

bkamins replied:
But do you test the case where an incorrect `GroupKey` is passed? It seems that that test must fail (`gd[key]` for `GroupKey` does not seem to throw `KeyError`). If you intend for it to work this way then maybe `GroupKey` should not be allowed in this definition?

jlumpe replied:
Removed `GroupKey` from the signature

bkamins commented:
I think we are almost there 😄. I left some minor, mostly stylistic comments.

Also you need to resolve merge conflicts (this should not be problematic - simply another PR also added tests at the end of the same file you change).

Finally - please make sure that in tests you test all functions that you define in all cases (I have not checked that yet, as we are working on API still, but this will be the thing that will have to be checked when we finalize the implementation).

In general I think that this PR is excellent - thank you again!

bkamins commented:
Thank you for the responses. I did not respond to them in most places because all responses would relate to the same thing, i.e. throwing errors. I had a feeling that the logic could be simplified.

E.g. look at this code:
```
function Base.getindex(gd::GroupedDataFrame, key::GroupKey)
    gd === parent(key) && return gd[getfield(key, :idx)]
    throw(ErrorException("Cannot use a GroupKey to index a GroupedDataFrame other than the one it was derived from."))
end

function Base.getindex(gd::GroupedDataFrame, key::Tuple)
    for (i, v) in enumerate(groupvalues(Tuple, gd))
        isequal(v, key) && return gd[i]
    end
    throw(KeyError(key))
end

function Base.getindex(gd::GroupedDataFrame, key::NamedTuple{N}) where {N}
    if length(key) != length(gd.cols) || any(n != _names(gd)[c] for (n, c) in zip(N, gd.cols))
        throw(KeyError(key))
    end
    return gd[Tuple(key)]
end
```

The first method throws `ErrorException` on error and the second and third throw `KeyError`. This will mean that `get` will behave differently for them, as it expects to intercept `KeyError` only (and the first method may throw `ErrorException` or `BoundsError`, but will never throw `KeyError`).

bkamins commented:
In summary - what I want to say is that writing code that correctly handles and intercepts exceptions is usually very tricky (as e.g. the same exceptions can be thrown by functions in Base, or opposite - some other exceptions can be thrown on the way and would not get caught).

Therefore it is crucial to carefully analyze the whole code in this respect (I mentioned one occurrence above that I was not sure generated the behavior you intended, but there might be more). Also it is usually best to try keeping the places where we throw exceptions to the minimum (if possible) as otherwise it very quickly becomes hard to analyze (exception is like a "goto" statement and generates similar problems - at least for me - in analyzing the logic of the code).

jlumpe commented:
The first one was actually intentional, I was thinking using a `GroupKey` from another `GroupedDataFrame` was a different sort of error than the others (and thus should propagate outside of `get`). But if you think treating it like the other cases is the best way to go I'll change it.

bkamins commented:
> But if you think treating it like the other cases is the best way to go I'll change it.

I am open to discussing it 😄. Simply it seems that `get` for `GroupKey` is not needed given the current design (it will never return `default`). So my intuition is - if that is intentional, then it is better not to allow `GroupKey` in `get`.

Then the logic of the whole PR would be: `GroupedDataFrame` behaves like a dictionary if passed `Tuple` or `NamedTuple` but behaves like a vector if passed `Integer` or `GroupKey`. Is this your intention? (I would be OK with this - probably it would be good to add one sentence to the documentation if this is so)

bkamins commented:
A second thing. Given this change maybe we should tighten the signature of:
```
function Base.getindex(gd::GroupedDataFrame, idxs::AbstractArray)
```
(it does not have to be in this PR)

The general question is if we want to allow the new options for single element `getindex` to be used in vector indexing.

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -65,6 +65,14 @@ See the following for additional split-apply-combine operations:
+`GroupedDataFrame` also supports the dictionary interface. The keys
+are the `GroupKey` objects returned by [`Base.keys(::GroupedDataFrame)`](@ref),
+which can also be used to get the values of the grouping columns for each group.
+`Tuples` and `NamedTuple`s containing the values of the grouping columns are
```
I'd say "in the same order as the `cols` argument" here, and drop the last sentence.

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -104,6 +112,14 @@ julia> last(gd)
 │ 1   │ 4     │ 1     │ 4     │
 │ 2   │ 4     │ 1     │ 8     │
 
+julia> gd[(a=3,)]
```
Can you also show a tuple and a `GroupKey`?

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
 Base.getindex(gd::GroupedDataFrame, idxs::Colon) =
     GroupedDataFrame(gd.parent, gd.cols, gd.groups, gd.idx, gd.starts, gd.ends)
 
+
```
```suggestion
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+
+
+"""
+    GroupKey
```
```suggestion
    GroupKey{T<:GroupedDataFrame}
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+
+See [`keys(::GroupedDataFrame)`](@ref) for more information.
+"""
+struct GroupKey{P<:GroupedDataFrame}
```
`T` is more common than `T` AFAIK.

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+end
+
+"""
+    groupvalues([T::Type], gd::GroupedDataFrame)
```
Do we still need this function? My idea was that `keys` (or `NamedTuple.(keys(gd))`) would replace it.

jlumpe replied:
That works for getting all the keys at once. There are also methods for getting values for a specific index and column, though, which I think could be useful for end users and are also used internally quite a bit.

nalimilan replied:
Well in that case I'd rather have `keys` return an indexable object (probably `<: AbstractVector`) so that you can extract the index for group `i`. Then you can also extract the column you want since it contains named tuples.

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+end
+
+function Base.show(io::IO, k::GroupKey)
+    print(io, "GroupKey")
```
```suggestion
    print(io, "GroupKey: ")
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+Base.length(key::GroupKey) = length(parent(key).cols)
+Base.names(key::GroupKey) = Tuple(groupvars(parent(key)))
+Base.keys(key::GroupKey) = names(key)
+Base.propertynames(::GroupKey) = names(key)
```
```suggestion
# Private fields are never exposed since they can conflict with column names
Base.propertynames(key::GroupKey, private::Bool=false) = names(key)
```
Also needs a test apparently. :-)

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+Base.propertynames(::GroupKey) = names(key)
+Base.values(key::GroupKey) = groupvalues(Tuple, parent(key), getfield(key, :idx))
+
+Base.iterate(key::GroupKey) = iterate(key, 1)
```
Remove this and put `i=1` in the next definition.

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+"""
+    keys(gd::GroupedDataFrame)
+
+Get the set of keys for each group in `GroupedDataFrame` object `gd`.
```
```suggestion
Get the set of keys for each group in `GroupedDataFrame` object `gd`
in the order in which groups appear.
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+│ 1   │ baz    │ 1     │ 6     │
+│ 2   │ baz    │ 1     │ 12    │
+
+julia> collect(keys(gd))
```
This doesn't sound like a good suggestion, better collect named tuples:
```suggestion
julia> NamedTuple.(keys(gd))
```

Also show the output of `keys(gd)` first.

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+ GroupKey(a = :bar, b = 2)
+ GroupKey(a = :baz, b = 1)
+
+julia> k = first(keys(gd));
```
Drop the semicolon so that the value is printed?

nalimilan commented on src/groupeddataframe/grouping.jl:
````diff
@@ -181,6 +197,319 @@ end
+│ 2   │ foo    │ 2     │ 7     │
+```
+"""
+function Base.keys(gd::GroupedDataFrame)
````
Turn this into a one-liner?

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -181,6 +197,319 @@ end
+
+Get a group based on the values of the grouping columns.
+
+`key` may be a `NamedTuple` or `Tuple` of grouping column values, or one of the
```
Mention that order must be respected.

nalimilan commented on test/grouping.jl:
```diff
@@ -1218,4 +1270,113 @@ end
     end
 end
 
+@testset "Dictionary interface" begin
```
```suggestion
@testset "GroupedDataFrame dictionary interface" begin
```

nalimilan commented:
> A second thing. Given this change maybe we should tighten the signature of:
> 
> ```
> function Base.getindex(gd::GroupedDataFrame, idxs::AbstractArray)
> ```
> 
> (it does not have to be in this PR)
> 
> The general question is if we want to allow the new options for single element `getindex` to be used in vector indexing.

In theory yes, in practice it might be tricky to handle, especially the case where values of different types may be mixed. So maybe simply throw an error if we encounter a value which isn't a symbol or integer in the index vector? Or maybe that's already what happens in practice, but then testing it would be nice.

jlumpe commented:
* Resolved all(?) remaining reviews
* Removed `T<:GroupKey` from `get(::GroupedDataFrame, key::T, default)`
* Tightened signature of `getindex(::GroupedDataFrame, ::AbstractArray)` to 
  `getindex(::GroupedDataFrame, ::AbstractArrray{T}) where {T<:Integer}`.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -177,6 +213,318 @@ end
+    groupvalues([T::Type], gd::GroupedDataFrame, i)
+    groupvalues(gd::GroupedDataFrame, i, col)
+
+Get the values of the grouping columns for `gd`, optionally specifying the group
```
maybe add a comment a generator is returned?

bkamins commented on src/groupeddataframe/grouping.jl:
````diff
@@ -177,6 +213,318 @@ end
+(:bar, 1)
+```
+"""
+groupvalues(T::Type, gd::GroupedDataFrame) = (groupvalues(T, gd, i) for i in 1:length(gd))
````
The consequence of being lazy here (generator) is that it will be slow as `groupvalues(T, gd, i)` has to perform repeatedly the same work for each `i`. Maybe let us return a `Vector` to make this faster? (then in the docs we will specify that we return a vector)? Do you have an opinion here?

jlumpe replied:
Not disagreeing that an array would be better here, but why would the performance be worse with the generator?

bkamins replied:
because `groupvalues(T, gd, i)` needs to determine the type of the element each time it is called. See e.g.:
```
NamedTuple{Tuple(groupvars(gd))}(groupvalues(Tuple, gd, i))
```
and you have to call `groupvars(gd)` in each iteration which adds some cost.

Also a generator does not work well with type inference (the caller will not know `eltype` even after using function barier).

Of course if we switched to array we have a tricky case that rows can have heterogeneous types so this means that the type of `Tuple` and `NamedTuple` will not be constant. And another level of complexity is that then we hit a distinction that `Tuple` is covariant and `NamedTuple` is invariant (but I guess we should default to standard Julia promotion rules here).

So - in general the case is not that easy to do it right. Therefore I do not push for this change very strongly. As generator is essentially less functional than a vector we can make this change later if we want to (but maybe you and @nalimilan come to the conclusion that it is better to do it in this PR).

nalimilan replied:
You mean a custom lazy `AbstractVector`? There are other advantages to that see [above thread](https://github.com/JuliaData/DataFrames.jl/pull/1908/files#r337931263). Regarding the computation of the element type, I guess we should just use the eltype of each column vector.

bkamins replied:
I would just use `Vector`.

> I guess we should just use the eltype of each column vector.

You mean for the `eltype` of the vector? - this should be OK for `Tuple` as it is covariant, but it will be problematic for `NamedTuple` because it is invariant, so the pushed `NamedTuples` would get `convert`ed to a supertype.

@jlumpe - sorry for this discussion (without clear conclusions and considering various options), but at least for me, this part of the design is tricky to get right.

nalimilan replied:
Well a `Vector` of names tuples is generally not what people need. Also that requires making a copy, which is probably not very useful. We could almost return a `StructArray`.

> You mean for the `eltype` of the vector? - this should be OK for `Tuple` as it is covariant, but it will be problematic for `NamedTuple` because it is invariant, so the pushed `NamedTuples` would get `convert`ed to a supertype.

Pushed to what? If the vector contains named tuples of the same type (the eltype of the columns), I don't see when the problem would happen.

bkamins replied:
If the `eltype` of column is `Integer` for example here you have what happens:
```
julia> v = NamedTuple{(:a,),Tuple{Integer}}[]
0-element Array{NamedTuple{(:a,),Tuple{Integer}},1}

julia> z = (a=1,)
(a = 1,)

julia> push!(v, z)
1-element Array{NamedTuple{(:a,),Tuple{Integer}},1}:
 NamedTuple{(:a,),Tuple{Integer}}((1,))

julia> v[1]
NamedTuple{(:a,),Tuple{Integer}}((1,))

julia> typeof(z)
NamedTuple{(:a,),Tuple{Int64}}
```
so you get a silent promotion.

bkamins replied:
I have written about `push!` but essentially it is an issue of storage:
```
julia> v = Vector{NamedTuple{(:a,),Tuple{Integer}}}(undef, 1)
1-element Array{NamedTuple{(:a,),Tuple{Integer}},1}:
 #undef

julia> v[1] = z
(a = 1,)

julia> v
1-element Array{NamedTuple{(:a,),Tuple{Integer}},1}:
 NamedTuple{(:a,),Tuple{Integer}}((1,))
```
is the same

bkamins replied:
in summary - maybe for the time being it is OK to stay with the generator and resolve if we want something more fancy in a separate PR. I am not sure what is best.

nalimilan replied:
But why would promotion be a problem? Anyway that vector would be read-only.

The reason why I suggest using a vector is that it would allow us to avoid adding `groupvalues`, and use `keys` directly instead.

bkamins replied:
So can you sumarize what you propose please? Is it to make `keys` return `Vector{GroupKey}` and then the user can just call `Tuple` or `NamedTuple` on it if one wants to get a specific type (similarly to what we currently have with `DataFrameRow`)? It should have the same performance as what `groupvalues` now has. Is this your idea. If yes I think it is OK.

nalimilan replied:
I was thinking of having `keys` return a custom `AbstractVector` which would lazily create `GroupKey` objects on indexing.

bkamins replied:
This is also OK (I understand that this custom `AbstractVector` would be also read-only). This should not be that hard (we do a similar thing e.g. for `eachrow`), we just have to remember to follow https://docs.julialang.org/en/latest/manual/interfaces/#man-interface-array-1 (which seems easy, but then usually some corner cases pop up 😄).

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -177,6 +213,318 @@ end
+The keys are [`GroupKey`](@ref) objects which behave like `NamedTuple`s holding
+the values of the grouping keys and can be used to index into `gd` efficiently.
+
+### Returns
```
do we need this `### Returns` section? Maybe just extend the first sentence of the docsting?

bkamins commented on src/groupeddataframe/grouping.jl:
````diff
@@ -177,6 +213,318 @@ end
+│ 2   │ foo    │ 2     │ 7     │
+```
+"""
+Base.keys(gd::GroupedDataFrame) = (GroupKey(gd, i) for i in 1:length(gd))
````
Here (as opposed to `groupvalues`) I think it is OK to be lazy (I am leaving this comment as maybe because we have a lazy option here we can choose to have an eager implementation of `groupvalues`).

bkamins commented:
Thank you for working on this. Apart from the inline comments I have left I have one general question.
If we are changing `getindex` of `GroupedDataFrame` we have to consider adding support for:
* `gd[AbstractArray{<:GroupKey}]`
* `gd[AbstractArray{<:Tuple}]`
* `gd[AbstractArray{<:NamedTuple}]`
* `gd[Not(selection)]`

The first three are introduced in this PR. We do not have to support them - again - I just want to be sure it is a conscious design decision.

The last one should be added to be consistent with the whole design of indexing we have, but it can be a separate PR.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -177,6 +213,318 @@ end
+optional first argument to the first two forms specifies the return type, which
+may be `NamedTuple` (default) or `Tuple`.
+
+### Returns
```
Same here - do we need `### Returns` section (please consider if we can handle it in the main body of the docstring), as this would shorten it which is desirable (if we do not loose precision).

Also - in the whole PR - can you please change `###` to `#` in docstrings for headings as this is a standard we use now in DataFrames.jl (and it is consistent with Base)?

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -61,6 +61,13 @@ See the following for additional split-apply-combine operations:
 * [`combine`](@ref) : combine a `GroupedDataFrame`, optionally applying a function to each group
 
+`GroupedDataFrame` also supports the dictionary interface. The keys are
+[`GroupKey`](@ref) objects returned by [`Base.keys(::GroupedDataFrame)`](@ref),
```
```suggestion
[`GroupKey`](@ref) objects returned by [`keys(::GroupedDataFrame)`](@ref),
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -177,6 +213,318 @@ end
+
+Base.parent(key::GroupKey) = getfield(key, :parent)
+Base.length(key::GroupKey) = length(parent(key).cols)
+Base.keys(key::GroupKey) = Tuple(groupvars(parent(key)))
```
Same here, why not just return the vector?
```suggestion
Base.keys(key::GroupKey) = groupvars(parent(key))
```

jlumpe replied:
This was for consistency with `NamedTuple`.

nalimilan replied:
Yeah, I realize we also do that for `DataFrameRow`, but I wonder whether there's any advantage to that.

@bkamins Do you have an opinion?

bkamins replied:
The only advantage is consistency with `NamedTuple` other than that there is no benefit. This is not a performance critical function so I think having it consistent with `NamedTuple` is OK.

jlumpe commented:
* Implemented lazy `GroupKeys <: AbstractVector{GroupKey}`, returned by `keys(::GroupedDataFrame)`.
* Fixed header levels in docstrings

If it's all right I would like to leave `getindex(::GroupedDataFrame, ::AbstractArray)` with non-integers for another PR, just so we can finally get this one merged.

The only thing left is `groupvalues`. I guess `GroupKeys` now provides nearly all of this functionality, and so could be demoted back to a private function.

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -177,6 +213,351 @@ end
+
+Base.parent(gk::GroupKeys) = gk.parent
+
+Base.length(gk::GroupKeys) = length(parent(gk))
```
a small stylistic comment - normally you would define `size` and `length` would use a default implementation. This is not a big thing though.

bkamins commented:
I agree we should try to finish this PR 😄. I would not expose `groupvalues` as `GroupKeys` should be enough (we can always expose it later if we find it really needed).

jlumpe commented:
Made `groupvalues` private again, removed alternate methods for selecting `Tuple`/`NamedTuple` return type and just return `DataFrameRow`.

bkamins commented on docs/src/lib/indexing.md:
```diff
@@ -197,3 +197,26 @@ Note that `sdf[!, col] .= v` and `sdf[!, cols] .= v` syntaxes are not allowed as
+The "elements" or "values" of a `GroupedDataFrame` are [`SubDataFrame`](@ref)s of its parent.
+
+* `gd[i::Integer]` -> Get the `i`th group.
+* `gd[a::AbstractArray{<:Integer}]` -> Get a reduced `GroupedDataFrame` containing only the groups
```
This should be a correct contract - right? If yes - can we have all these options in the tests. Thank you!
```suggestion
* `gd[a::Union{AbstractVector{<:Integer}, Colon, Not}]` -> Get a reduced `GroupedDataFrame` containing only the groups
```

jlumpe replied:
This behavior hasn't changed in this PR (aside from restricting the signature), so I'd rather update this in a follow-up PR if that's alright.

bkamins replied:
OK

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -156,10 +190,12 @@ Compat.lastindex(gd::GroupedDataFrame) = length(gd.starts)
 
-function Base.getindex(gd::GroupedDataFrame, idxs::AbstractArray)
+# Array of integers
+function Base.getindex(gd::GroupedDataFrame, idxs::AbstractArray{T}) where {T<:Integer}
```
following my comment above
```suggestion
function Base.getindex(gd::GroupedDataFrame, idxs::Union{AbstractVector{T}, Not}) where {T<:Integer}
```

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -177,6 +213,292 @@ end
+
+Base.size(gk::GroupKeys) = (length(parent(gk)),)
+Base.IndexStyle(::GroupKeys) = IndexLinear()
+function Base.getindex(gk::GroupKeys, i::Integer)
```
I think this is needed to make `@bounschck` work:
```suggestion
Base.@propagate_inbounds function Base.getindex(gk::GroupKeys, i::Integer)
```

bkamins commented:
Looks good except for the small comments I have made.

I will go through the test coverage of the PR in detail later after we have finalized the tiny design things I have raised (but please make sure that we test all corner cases - maybe this is already so, but I have not had time to check it in detail yet and I want to move forward to merge it as soon as possible to include it in v0.20 release 😄).

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -177,6 +213,292 @@ end
+"""
+    GroupKeys{T<:GroupedDataFrame} <: AbstractVector{GroupKey{T}}
+
+An abstract vector containing all [`GroupKey`](@ref) objects for a given
```
```suggestion
A vector containing all [`GroupKey`](@ref) objects for a given
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -177,6 +213,292 @@ end
+Base.parent(gk::GroupKeys) = gk.parent
+
+Base.size(gk::GroupKeys) = (length(parent(gk)),)
+Base.IndexStyle(::GroupKeys) = IndexLinear()
```
This is more general as it allows calling the function with only the type, even if you don't have an instance:
```suggestion
Base.IndexStyle(::Type{<:GroupKeys}) = IndexLinear()
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -177,6 +213,292 @@ end
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 2     │ 7     │
+
+julia> isequal(gd[keys(gd)[1]], gd[1])
```
Just for simplicity:
```suggestion
julia> gd[keys(gd)[1] == gd[1]
```

nalimilan commented on docs/src/lib/indexing.md:
```diff
@@ -197,3 +197,26 @@ Note that `sdf[!, col] .= v` and `sdf[!, cols] .= v` syntaxes are not allowed as
+returned by [`keys(::GroupedDataFrame)`](@ref), which behaves similarly to a `NamedTuple` but has
+performance on par with integer indexing.
+
+The "elements" or "values" of a `GroupedDataFrame` are [`SubDataFrame`](@ref)s of its parent.
```
```suggestion
The elements of a `GroupedDataFrame` are [`SubDataFrame`](@ref)s of its parent.
```

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -238,76 +238,15 @@ function _groupvar_idx(gd::GroupedDataFrame, name::Symbol, strict::Bool)
+_groupvalues(gd::GroupedDataFrame, i::Integer, col::Symbol) = _groupvalues(gd, i, _groupvar_idx(gd, col, true))
 
 # Eltypes of the grouping columns
 _grouptypes(gd::GroupedDataFrame) = Tuple(eltype(gd.parent[!, c]) for c in gd.cols)
```
This isn't used anymore it seems.

jlumpe replied:
This are used in `getindex` and `getproperty` of `GroupKey`, I think.

bkamins replied:
I do not see the usage of this function in this PR.

jlumpe replied:
Oh, I though this was about `_groupvalues` methods. I've removed `_grouptypes`.

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -177,6 +213,292 @@ end
+Base.names(key::GroupKey) = groupvars(parent(key))
+# Private fields are never exposed since they can conflict with column names
+Base.propertynames(key::GroupKey, private::Bool=false) = keys(key)
+Base.values(key::GroupKey) = Tuple(_groupvalues(parent(key), getfield(key, :idx)))
```
Just like for `names`:
```suggestion
Base.values(key::GroupKey) = _groupvalues(parent(key), getfield(key, :idx))
```

jlumpe replied:
I had `keys` and `values` return `Tuple`s to match `NamedTuple`.

nalimilan replied:
Cf. discussion in the other thread: https://github.com/JuliaData/DataFrames.jl/pull/1908/files/04bfed58fb6db9749b94eadb372439cb1e184958#r350175810 Let's keep that then...

bkamins commented:
@jlumpe - what is left to be done in this PR (there are some unresolved conversations, but maybe they are outdated or something more (like additional tests or docs update) need to be done - I guess you know best 😄).

I would like to go with 0.20 release that would include this PR.

jlumpe commented:
OK, full summary of the PR:

* Implemented dictionary interface for `GroupedDataFrame`:
  * `keys(gd)` returns `GroupKeys`, lazy vector of `GroupKey` objects
    * `GroupKey` objects behave like `NamedTuple`s of grouping column values, but property/element access is lazy. Allows for retrieval of grouping column values for each group.
  * `getindex(gd, ...)` defined for dictionary keys:
    * `GroupKey` objects belonging to the `GroupedDataFrame`.
    * `Tuple` or `NamedTuple` of grouping column values.
  * `get(gd, ...)` for `Tuple` and `NamedTuple`, not for `GroupKey` because they are always valid keys (for the correct `GroupedDataFrame`).
* Added section on `GroupedDataFrame`s to indexing page in documentation
  * Clarify difference between array-style and dictionary-style indexing
  * List of all valid argument types for `getindex`.
* Refined signature of `getindex(gd, ::Array)` to restrict it to integer arrays, which was the existing contract of the method.
* Added additional tests for `groupvars` and `groupindices` with multiple grouping columns.
* Added test set for column names changing after `groupby` called.

I think this is all pretty complete. There are a couple of reviews that it won't let me mark as resolved for some reason. I'm going to go over it once more and check for any corner cases left out of the tests, but I think it should be ready to merge.

bkamins commented on docs/src/lib/indexing.md:
```diff
@@ -197,3 +197,26 @@ Note that `sdf[!, col] .= v` and `sdf[!, cols] .= v` syntaxes are not allowed as
+  grouping columns. The fields of the `NamedTuple` must match the grouping columns
+  columns passed to [`groupby`](@ref) (including order).
+* `gd[key::Tuple]` -> Same as previous, but omitting the names on `key`.
+* `get(gd, key, default)` -> Get group by (named) Tuple `key`, returning `default` if it does not exist.
```
I think it is clearer to be explicit
```suggestion
* `get(gd, key, default)` -> Get group by `NamedTuple` or `Tuple` `key`, returning `default` if it does not exist.
```

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -99,6 +106,33 @@ julia> last(gd)
+├─────┼───────┼───────┼───────┤
+│ 1   │ 3     │ 2     │ 3     │
+│ 2   │ 3     │ 2     │ 7     │
+
```
maybe add one example with array indexing (returning a `GroupedDataFrame`)?

jlumpe replied:
I'll probably do this in the follow-up, which expands array indexing.

bkamins replied:
OK

bkamins commented on src/groupeddataframe/grouping.jl:
```diff
@@ -177,6 +213,292 @@ end
+    end
+end
+
+function Base.NamedTuple(key::GroupKey)
```
Maybe add the note of existence of conversion methods in the documentation of `GroupKey`?

jlumpe replied:
This is contained in the docstring to `GroupKeys`.

bkamins replied:
You only show `NamedTuple(k)` in the example if I understand things correctly. I think it should be added to the description of the type (as in this case it is actually relevant that these conversions exist).

We can add it in the second PR regarding this functionality which I understand you are planning to submit.

jlumpe replied:
Added as a comment as the list of examples is already fairly long:
```
julia> values(k)  # Can also use Tuple(k)
(:foo, 2)
```

bkamins replied:
ok

bkamins commented on test/grouping.jl:
```diff
@@ -1256,4 +1276,131 @@ end
+    df = DataFrame(a = repeat([:A, :B, missing], outer=4), b = repeat([:X, :Y], inner=6), c = 1:12)
+    gd = groupby_checked(df, [:a, :b])
+
+    @test map(NamedTuple, keys(gd)) ≅
```
do we have the same set of tests for `Tuple`?

jlumpe replied:
Those are below in the "GroupKey and GroupKeys" test set. This is just to check the keys are what I think they are before I check each one individually.

bkamins replied:
OK

bkamins commented on test/grouping.jl:
```diff
@@ -1256,4 +1276,131 @@ end
+    @test_throws KeyError gd[(b=:X, a=:D)]
+end
+
+@testset "GroupKey and GroupKeys" begin
```
do we have tests for invalid keys?

jlumpe replied:
What do you mean by an invalid key? I test indexing with `Tuple`s/`NamedTuple`s with incorrect values, incorrect/out of order names, and too many or too few elements, all of which throw a `KeyError`. I also test for using a `GroupKey` from a `GroupedDataFrame` which is equal to but not the same object as the one being indexed. Anything else is a `MethodError`.

bkamins replied:
I meant testing indexing of `GroupKeys` as this is also mentioned in the `testset` name (i.e. that it is only allowed to index them within bounds). This is a trivial test, but it is good to have it (maybe we have it but I missed it), because in the future if someone changes the implementation such trivial things can break.

jlumpe replied:
Oh, you mean with integer indices? Yes, that should be added.

bkamins replied:
Yes - thank you!

bkamins commented on test/grouping.jl:
```diff
@@ -1256,4 +1276,131 @@ end
+    @test NamedTuple(keys(gd)[1]) == (a=:A, b=:X)
+    @test keys(gd)[1].a == :A
+
+    names!(df, [:d, :e, :f])
```
`names!` is deprecated - use `rename!`

bkamins replied:
Why is this resolved? I do not see the change in the code.

jlumpe replied:
Sorry, I mark these off preemptively as I fix them in my local repo. Usually push the changes soon after but haven't yet this time.

bkamins commented:
Looks good. I left a few minor comments and we can merge it.

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -177,6 +213,292 @@ end
+Base.parent(gk::GroupKeys) = gk.parent
+
+Base.size(gk::GroupKeys) = (length(parent(gk)),)
+Base.IndexStyle(::Type{GroupKeys}) = IndexLinear()
```
You didn't apply my suggestion exactly. The `<:` is really needed.
```suggestion
Base.IndexStyle(::Type{<:GroupKeys}) = IndexLinear()
```
(BTW, please don't squash commits as it makes it painful to spot what changed.)

jlumpe replied:
Why is it needed? `GroupKeys` is a `DataType`/leaf type so `<:GroupKeys` shouldn't match anything but `GroupKeys` itself, right?

bkamins replied:
This is due a standard problem with parametric types, e.g.:
```
julia> Vector{Int} isa Type{Vector}
false

julia> Vector{Int} isa Type{<:Vector}
true
```

jlumpe replied:
Aha, you're absolutely right. Fixed.

jlumpe commented:
Ok, think that should resolve all remaining issues.

bkamins commented:
The only thing left is `_grouptypes` function definition that seems not used so I think it should be removed. See https://github.com/JuliaData/DataFrames.jl/pull/1908#discussion_r353375306 comment

nalimilan commented on src/groupeddataframe/grouping.jl:
```diff
@@ -177,6 +213,289 @@ end
+end
+
+# Get values of grouping columns for all groups
+_groupvalues(gd::GroupedDataFrame) = [_groupvalues(gd, i) for i in 1:length(gd)]
```
This one isn't used either AFAICT.

jlumpe replied:
Removed.

nalimilan commented:
There are a few uncovered lines according to Coveralls. Can you check whether they need tests? See https://coveralls.io/builds/27423710/source?filename=src%2Fgroupeddataframe%2Fgrouping.jl

bkamins commented:
The display tests are usually put in test/show.jl.
The test of `IndexStyle` is used by `pairs` by default.

bkamins commented:
@jlumpe - can you please have a look at comments by @nalimilan. I want to have this PR i 0.20 and this is the last one scheduled for the release. Thank you.

jlumpe commented:
@bkamins Can you clarify what you mean about `IndexStyle`?

bkamins commented:
> Can you clarify what you mean about `IndexStyle`?

`IndexStyle` is now not covered by tests.

In order to include tests of `IndexStyle` you either have to call this function directly or call `pairs` as it internally calls `IndexStyle`. This is a minor issue so I think we can leave this out.

@nalimilan - if you will not have additional comments I will merge this PR today.

bkamins commented:
Ah - this is a very minor issue so I am OK to merge this PR without it (just please add it in the next PR you are planning to do related to this functionality).

bkamins commented on docs/src/lib/functions.md:
```diff
@@ -16,7 +16,10 @@ by
 combine
 groupby
 groupindices
+groupvalues
```
```suggestion
```

bkamins replied:
I do not know how to force-push this. I will remove this line in 0.20 PR as this function does not exist now
